File: e2e/qwik-e2e/apps/vdomless-counter/src/build-data.rs
//! Native implementation of `./build-data.ts` (specs/09 user plugin): same semantics —
//! LCG randomness, process-global id counter, rows of nested signals.

use qwik::native::{IntoSerdes, Signal};
use qwik::serdes::SerdesValue;
use std::rc::Rc;
use std::sync::atomic::{AtomicU64, Ordering};

const ADJECTIVES: &[&str] = &[
	"pretty", "large", "big", "small", "tall", "short", "long", "handsome", "plain", "quaint",
	"clean", "elegant", "easy", "angry", "crazy", "helpful", "mushy", "odd", "unsightly",
	"adorable", "important", "inexpensive", "cheap", "expensive", "fancy",
];
const COLORS: &[&str] = &[
	"red", "yellow", "blue", "green", "pink", "brown", "purple", "brown", "white", "black",
	"orange",
];
const NOUNS: &[&str] = &[
	"table", "chair", "house", "bbq", "desk", "car", "pony", "cookie", "sandwich", "burger",
	"pizza", "mouse", "keyboard",
];

static NEXT_ID: AtomicU64 = AtomicU64::new(1);

pub struct Row {
	id: u64,
	label: Signal<String>,
	selected: Signal<bool>,
}

impl IntoSerdes for Row {
	fn into_serdes(self) -> Rc<SerdesValue> {
		Rc::new(SerdesValue::Object(vec![
			("id".to_string(), self.id.into_serdes()),
			("label".to_string(), self.label.into_serdes()),
			("selected".to_string(), self.selected.into_serdes()),
		]))
	}
}

/// A spliced sidecar cannot pull in `rand`, so entropy comes from std's per-process hasher keys.
fn random(max: usize) -> usize {
	use std::collections::hash_map::RandomState;
	use std::hash::{BuildHasher, Hasher};
	(RandomState::new().build_hasher().finish() as usize) % max
}

pub fn buildData(count: usize) -> Vec<Row> {
	let mut rows = Vec::with_capacity(count);
	for _ in 0..count {
		let label = format!(
			"{} {} {}",
			ADJECTIVES[random(ADJECTIVES.len())],
			COLORS[random(COLORS.len())],
			NOUNS[random(NOUNS.len())]
		);
		rows.push(Row {
			id: NEXT_ID.fetch_add(1, Ordering::Relaxed),
			label: Signal::new(label),
			selected: Signal::new(false),
		});
	}
	rows
}

File: e2e/qwik-e2e/apps/vdomless-counter/src/build-data/build-data.ts
import { native$, nativeFrom, useSignal } from '@qwik.dev/core';

const adjectives = ["pretty", "large", "big", "small", "tall", "short", "long", "handsome", "plain", "quaint", "clean", "elegant", "easy", "angry", "crazy", "helpful", "mushy", "odd", "unsightly", "adorable", "important", "inexpensive", "cheap", "expensive", "fancy"]; // prettier-ignore
const colors = ["red", "yellow", "blue", "green", "pink", "brown", "purple", "brown", "white", "black", "orange"]; // prettier-ignore
const nouns = ["table", "chair", "house", "bbq", "desk", "car", "pony", "cookie", "sandwich", "burger", "pizza", "mouse", "keyboard"]; // prettier-ignore

const random = (max: number) => Math.round(Math.random() * 1000) % max;

let nextId = 1;

type Signal<T> = { value: T };

export type Row = {
  id: number;
  label: Signal<string>;
  selected: Signal<boolean>;
};

export const buildData = native$(
  (count: number): Row[] => {
    const data = new Array(count);
    for (let i = 0; i < count; i++) {
      const label = useSignal(
        `${adjectives[random(adjectives.length)]} ${colors[random(colors.length)]} ${nouns[random(nouns.length)]}`
      );
      data[i] = {
        id: nextId++,
        label,
        selected: useSignal(false),
      };
    }
    return data;
  },
  { rust: nativeFrom('./native') }
);
